bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/AwattarBridgeHandler.java

```diff
@@ -240,7 +240,7 @@ public synchronized SortedSet<AwattarPrice> getPrices() {
     public boolean containsPriceFor(long timestamp) {
         SortedSet<AwattarPrice> localPrices = getPrices();
         return localPrices != null && localPrices.first().timerange().start() <= timestamp
-                && localPrices.last().timerange().end() > timestamp;
+                && localPrices.last().timerange().end() >= timestamp;
```

Comment:
This is fishy. For a period with time, **end** is considered **exclusive**/closed. As an example, these periods:
- 01.11.2024 23:00:00 - 02.11.2024 00:00:00
- 02.11.2024 00:00:00 - 02.11.2024 01:00:00

do not overlap, even though end time of first period equals start time of second period.

When checking which period the timestamp 02.11.2024 00:00:00 belongs to, it would be the second one, since the first one ends _just before_ that.

So the real question is which timestamps are provided by the API? Are they like in the example above, or are they actually provided like this:
- 01.11.2024 23:00:00 - 01.11.2024 23:59:59
- 02.11.2024 00:00:00 - 02.11.2024 01:00:00

?

If this is the case, there are other problems in the binding as well, since such periods are not strictly correct and would need to be interpreted differently elsewhere. For example in the `TimeRange` method mentioned below.

The current logic can be simplified to:
```java
return localPrices != null && localPrices.first().timerange().contains(timestamp);
```

Reply:
Good point, need to re-investigate. As it only seems to occur at midnight, it might be that only the last hour of the day is at 59:59, but before i do some more guessing best is to get some data logged. Unfortunately i'll need to ask on the forums as we don;t have awattar here.

Reply:
The API does Indeed return start and end times to the full hour. 

```
    {
      "start_timestamp": 1731265200000,
      "end_timestamp": 1731268800000,
      "marketprice": 125.31,
      "unit": "Eur/MWh"
    },
```

Reply:
First, let me correct myself regarding the "simplification" I mentioned above. Should be:
```java
        if (localPrices == null) {
            return false;
        }
        return new TimeRange(localPrices.first().timerange().start(), localPrices.last().timerange().end())
                .contains(timestamp);
    }
```

whether that's a simplification or not is up to the reader. 😄 But it would require constructing a new `TimeRange` record, so for sure it's not a performance optimization.

Anyway... since `end_timestamp` in this case is 2024-11-10T20:00:00+00:00 (as expected), we need to treat it as exclusive, and therefore the current implementation of `containsPriceFor` is correct, and the fix was wrong.

Reply:
yeah yeah yeah, as usual... 💯 

The issue remains to be there, the last hour is not selected. See https://github.com/openhab/openhab-addons/issues/17316#issuecomment-2442971546 (let's continue there)

Reply:
I'm not really into this binding, but maybe something is wrong here?
https://github.com/openhab/openhab-addons/blob/45cfd0335b7eee651eabbabdcd1b93f5c359badb/bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/AwattarBestPriceHandler.java#L144-L145

I guess `timerange.end()` could be e,g, 2024-11-10T20:00:00Z, which would not be within the range 2024-11-10T19:00:00Z - 2024-11-10T20:00:00Z? I guess the check here is to see if we have both prices for start and end, and if not, return. But since the end of this `TimeRange` is also exclusive, we don't need to have a price for 2024-11-10T20:00:00Z.

`containsPriceFor` doesn't have the context to know if the timestamp being asked for is start (inclusive) or end (exclusive).

The quick fix would then be what you both have suggested, but just in this place rather than in `containsPriceFor`. I'm not suggesting this as final solution, but you could start to see if this works:
```java
        if (!(bridgeHandler.containsPriceFor(timerange.start()) && bridgeHandler.containsPriceFor(timerange.end() - 1))) {
```

Reply:
> I guess timerange.end() could be e,g, 2024-11-10T20:00:00Z, which would not be within the range 2024-11-10T19:00:00Z - 2024-11-10T20:00:00Z?

Theoretical yes, but in the current implementation of the binding and the API this would never happen. 

Any time range in the current situation will be from XX:00:00 to XX:00:00 the next hour and will always be 1h long. 

So whenever you see two time ranges having the same start or end date, they will be technically the same. 

But again, this is the current situation of the binding and the API. But generally I doubt the APi will change anytime soon.

Reply:
I'll propose a PR today, with adapting the API.

I think generally the better solution is to not have the time ranges overlap at the XX:00:00 point.

I think the current situation is formally not correct, since time ranges, from my understanding, are strictly inclusive.

Therefore a 1h time range should last exactly 60 minutes, eg. 01:00:00 to 01:59:59. The last second is implicitly inclusive.

Reply:
Proposed here: #17731